### ISSUE TITLE
Add cirq.sample_final_wavefunction method

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#1740]
When a circuit has noise, it does not have a well defined final wavefunction. But as long as the noise is all in the form of mixtures, it is still possible to sample from the possible final wavefunctions as efficiently as if you were running a unitary simulation.

Things this issue involves:

- Update cirq.final_wavefunction error message to mention the sample_final_wavefunction as an alternative that can be used.
- Basically copy-paste cirq.final_wavefunction, but tweak it to allow mixtures but not fully general channels.

**Comment by vtomole** [comment]
Good first issue for the ambitious.

**Comment by Lucaman99** [comment]
I'm no expert, but I would like to try to give one of these issues a try. How difficult do you think this will be to resolve?

**Comment by vtomole** [comment]
@Lucaman99  I encourage you to try it out and let us know when you get stuck (if you do). Note: there are other easier issues here: https://github.com/quantumlib/Cirq/labels/good%20first%20issue

**Comment by Lucaman99** [comment]
@vtomole Great, I'll give it a shot! Thank you!

**Comment by Strilanc** [comment]
`cirq.Simulator` will already do the right thing if given a circuit with mixture gates, so there's no need to write new simulation code. It's mostly a matter of short-circuiting the simulation quickly when it's going to fail anyways.

**Comment by Lucaman99** [comment]
@Strilanc Awesome, I'll try doing that

**Comment by Lucaman99** [comment]
Hey @vtomole @Strilanc, sorry, this is irrelevant to the actual issue, but I've been trying to setup the virtual environment for Cirq development for a few hours now and for some the reason the virtualenvwrapper library is just not working. It doesn't recognize commands when I try to create a new virtual environment. Would you happen to have any suggestions for resolving this issue?

**Comment by Strilanc** [comment]
Sometimes you need to open a new terminal after installing.

virtualenvwrapper is a system dependency, so pip installing cirq or its python dependencies won't install it.

**Comment by Lucaman99** [comment]
@Strilanc Yeah I installed all of the requirements.txt and restarted the terminal, still not recognizing the command.

**Comment by Strilanc** [comment]
In that case I'm not sure what to do. You can always fall back to using virtual envs in the standard way, with a `.venv` folder somewhere.

**Comment by Lucaman99** [comment]
Ok, I'll give that a try, thanks!

**Comment by Lucaman99** [comment]
Hey @Strilanc, I've been playing around with the cirq.final_wavefunction method and it is making sense. I was wondering though, when you say you don't want this new method to allow "fully general channels", what exactly do you mean by that?

**Comment by Lucaman99** [comment]
Hey @Strilanc - So to construct the cirq.sample_final_wavefunction I took the cirq.final_wavefunction method and removed the condition that all gates have to be unitary. I tested it with a bit flip method acting on a qubit in a small circuit, and it seemed to work (it outputted the correct wavefunction). The issue is that I can only call this new method as cirq.sim.mux.sample_final_wavefunction. If I simply try to call cirq.sample_final_wavefunction, it doesn't work. Do you know why that would be happening?

**Comment by Lucaman99** [comment]
Also @Strilanc, how would one differentiate between mixtures and fully-general channels?

**Comment by Lucaman99** [comment]
In addition, I tried using the cirq.has_mixture as opposed to cirq.has_unitary method, but for some reason, this didn't work (it kept throwing the same error, meaning that cirq.has_mixture(arg) = False for an argument that should return True).

**Comment by Strilanc** [comment]
Hey @Lucaman99 ,

> when you say you don't want this new method to allow "fully general channels", what exactly do you mean by that?

A fully general channel is an operation that implements `_channel_` but not `_mixture_` or `_unitary_`. That is to say, it does not have an interpretation as a probabilistic set of unitaries acting only on the system qubits. @dabacon do you know a specific concrete example of such a channel?

> can only call this new method as cirq.sim.mux.sample_final_wavefunction

You have to import it within `cirq/sim/__init__.py` and then `cirq/__init__.py`.

> I tried using the cirq.has_mixture as opposed to cirq.has_unitary method, but for some reason, this didn't work (it kept throwing the same error, meaning that cirq.has_mixture(arg) = False for an argument that should return True).

This is likely a bug in `has_mixture`. What's the specific operation that's failing?

**Comment by Lucaman99** [comment]
Hey @Strilanc 

Thank you for your response! For the `has_mixture` method, it's essentially just failing when I try to replace the `has_unitary` with it. I played around with the function a bit more, and created a few simple circuits with some mixer gates (bit flip, phase damp, amplitude damp, etc.) and the `has_mixture` method seems to always return false, regardless of whether the gates in the circuit are unitary, mixture, etc.

**Comment by Lucaman99** [comment]
On the other hand, the `has_unitary` method is working exactly as expected, which leads me to believe that this is in fact a bug. Should I open an issue?

**Comment by Strilanc** [comment]
Yes, please open an issue. That's a serious bug.

**Comment by Lucaman99** [comment]
@Strilanc Will do! Any idea what could be causing the bug?

**Comment by Strilanc** [comment]
Probably something silly like checking for `_has_mixture_` but not `_has_unitary_` (which implies that a mixture can be derived).

**Comment by jkc1113** [comment]
I can try this out

**Comment by dabacon** [comment]
An example of a channel to test mixtures with can be created using `cirq.bit_flip`.

**Comment by satw1knandala** [comment]
Is help still needed on this? I see that the final_wavefunction is being deprecated for final_state_vector.

**Comment by balopat** [comment]
Yes @satw1knandala, we'd love to see this implemented, yes, I renamed the issue to state_vector.

**Comment by satw1knandala** [comment]
Sure, I will get started on this one. You can assign me. Thanks!

**Comment by satw1knandala** [comment]
update: still working on this. should be done by end of week

**Comment by satw1knandala** [comment]
@Strilanc @balopat i was able to create the function sample_final_state_vector, but I am writing tests and have a few questions - 

1. to test against a fully general channel, can i just create a random non unitary matrix and test it on that? or do you have a better solution to this?
2. should i replicate the tests for final_state_vector to sample_final_state_vector in mux_tests.py? the functions only really differ in one line of code (checking for mixture vs unitary) so I'm wondering if it might be repetitive?

Thanks!

**Comment by satw1knandala** [comment]
@balopat are we still tracking this? thank you

**Comment by daxfohl** [comment]
@artvandalay404 Feel free to post a PR with what you have. My knee jerk reaction is that similar test should be fine, but would have to see the actual code. Maybe there's a middle ground. Testing against a channel, I'd just use an existing channel like phase damp. I don't see a need to make a custom matrix.

**Comment by daxfohl** [comment]
@Strilanc Does this issue mean (a) sample a result efficiently from the union of possible final state vectors, or (b) run a simulation to get a single sample final state vector in its entirety?
* I _think_ it's (b), but if so, why does it need to be limited to mixtures? The state vector simulator can work with channels as well. Maybe this was not the case in 2019?
* If (a), the mixture restriction makes sense, as you can calculate a set of possible final state vectors and sample from them efficiently. Though the set will grow exponentially with each noise gate and be impractical for all but the shortest of circuits.

**Comment by matti-sarjala** [comment]
A useful way to keep this feature reviewable might be to define the allowed operation classes explicitly:

1. **Allowed**
   - unitary operations,
   - mixture operations, e.g. `cirq.bit_flip(p)`,
   - circuits where the simulator can sample one branch of the mixture and return the corresponding state vector.

2. **Rejected**
   - operations that only implement a general channel but do not have a unitary/mixture interpretation.
   - These should fail early with an error explaining that no sampled pure-state trajectory is available.

3. **Tests**
   - unitary-only circuit agrees with `final_state_vector`,
   - simple mixture circuit such as `cirq.bit_flip(p)` returns one valid sampled branch,
   - fully general channel raises a clear error,
   - `final_state_vector` error message points users to `sample_final_state_vector` when the circuit contains mixtures.

This would make the method’s contract clear: it samples pure-state trajectories through probabilistic unitary branches, but it does not simulate arbitrary noisy channels as state vectors.

**Comment by matti-sarjala** [comment]
A small toy-model distinction might help make the API boundary easier to test and document:

- A unitary circuit has one well-defined final state vector.
- A circuit with mixture noise does not have one deterministic final state vector, but it does have sampleable pure-state branches.
- A circuit with a fully general channel may require density-matrix semantics, because there may be no system-only pure-state branch corresponding to the channel.

A minimal example could be:

1. No noise:
   - final state vector is deterministic.

2. Bit flip noise:
   - with probability `p`, apply `X`;
   - with probability `1-p`, apply `I`;
   - `sample_final_state_vector` returns one branch state vector.

3. General channel:
   - reject with a clear error,
   - because the operation is not representable as a probabilistic choice of unitaries on the system alone.

This would make the feature boundary clear: the method is not “final_state_vector but allowing all noise”. It is “sample a pure-state trajectory when the noisy operation has mixture semantics”.
